feat(eslint-config-fluid): Add eslint rule to prevent imports of @internal items

WayneFerrao opened the pull request:
[AB#5695](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5695)


## Description

Fluid Framework needs to maintain a clear boundary between public and internal APIs. The usage of @internal and similar annotations is crucial for this, but we lacked automated enforcement. This PR adds a safeguard through an eslint rule that ensures only the intended, stable, and documented parts of the library's API are used in public-facing code samples and tests. There aren't any custom rules so new infra was added.

When implemented, this eslint rule would rely on including the necessary tags that we want to prevent the import of, ie: @internal, @alpha etc.
The exceptions property can be used to allow for internal imports from a specific package.



### Next Steps/ To DO
Update the README to include instructions for adding and testing new rules.

WayneFerrao committed:
WIP: added eslint rule to prevent imports of @internal items

WayneFerrao committed:
Add comments

WayneFerrao committed:
Update tests

WayneFerrao committed:
Added exception test

WayneFerrao committed:
Merge branch 'main' into addDevtoolsExistenceCheck

WayneFerrao committed:
Merge branch 'main' into addDevtoolsExistenceCheck

WayneFerrao commented:
Original PR closed because I had pushed the wrong branch. See previous discussion on this PR [here](https://github.com/microsoft/FluidFramework/pull/18259)

WayneFerrao committed:
Moved doctrine installation to package from root

WayneFerrao committed:
Merge branch 'main' into preventImportsRule

WayneFerrao commented:
@Josmithr I looked into using TSDoc parser. From what I can tell, it is feasible, the parser is designed to work with the AST generated by TypeScript. ESLint rules typically operate on the ESTree AST provided by ESLint, which is different from the TypeScript AST. I would need to import and parse the entire file, which would significantly impact performance and slow down linting. What do you think?

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,77 @@
+		type: "problem",
+		docs: {
+			description:
+				"This rule restricts imports from internal or non-public APIs. This to prevent accidental dependencies on internal, unstable or undocumented parts of the codebase.",
```
We'll want to update this to be a bit more general, since this rule will work for any tag.

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,77 @@
+							if (restrictedTags.has(tag.title)) {
+								context.report({
+									node,
+									messageId: "restrictedImport",
```
How about "importWithRestrictedTag"?

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,77 @@
+			category: "Best Practices",
+		},
+		fixable: "code",
+		schema: [
```
Is there any way for us to run a 1-time check on all the specified tags to ensure they're of the right form (e.g. starts with `@`)?

Josmithr replied:
Or if we can specify this declaratively in the schema somehow, but I'm guessing that isn't an option?

WayneFerrao replied:
Yeah, you can't specify it in the schema, but I added a check when pulling in the tags that throws an error if it isn't formatted appropriately.

WayneFerrao committed:
Removed doctrine, added tsdoc and updated tests

WayneFerrao committed:
updated comments

WayneFerrao committed:
Added comments for examples

WayneFerrao committed:
Merge branch 'main' into preventImportsRule

WayneFerrao committed:
Fixed policy check error

WayneFerrao committed:
Fix policy check error

alexvy86 commented on common/build/eslint-config-fluid/src/test/enforce-no-restricted-tags-imports.test.js:
```diff
@@ -0,0 +1,79 @@
+	valid: [
+		{
+			code: validCodeWithJSDoc,
+			// Valid import
```
The comment feels like it wants to describe the test, but if so I think it makes more sense to move it to above the full object inside the `valid` array (same for the ones below). And how about making it more explicit? Something like `Importing something with a non-restricted tag`.

alexvy86 commented on common/build/eslint-config-fluid/src/test/enforce-no-restricted-tags-imports.test.js:
```diff
@@ -0,0 +1,79 @@
+			options: [
+				{
+					tags: ["@internal", "@alpha"],
+					exceptions: ["./foo", "bar"],
```
The exceptions don't seem to be relevant for this test. The simpler it is, the better :).
```suggestion
```

alexvy86 commented on common/build/eslint-config-fluid/src/test/enforce-no-restricted-tags-imports.test.js:
```diff
@@ -0,0 +1,79 @@
+	],
+});
+
+console.log("All tests passed!");
```
So `ruleTester.run()` will error and exit if something fails and we won't hit this line?

WayneFerrao replied:
Yes, this log won't run if something fails.

alexvy86 commented on common/build/eslint-config-fluid/src/test/enforce-no-restricted-tags-imports.test.js:
I'm not sure the test are actually validating what they're supposed to... In all cases, the function we're putting the tag on is not the one being imported. We'd need to somehow simulate the existence of two files; one with the function definition and the tag, and a separate one that imports from the first one.

Also, is there a way to test what should happen when the configuration is invalid (e.g. a tag doesn't start with `@`)?

WayneFerrao replied:
@alexvy86 I don't think ESLint's RuleTester can simulate multiple files perfectly because it's meant to test rules on a per-file basis. At best, you could split the code into different strings/"files" and concatenate them to kinda simulate multiple files.

WayneFerrao replied:
And yes, I can add a test for invalid configs.

alexvy86 replied:
> @alexvy86 Alejandro Villarreal FTE I don't think ESLint's RuleTester can simulate multiple files perfectly because it's meant to test rules on a per-file basis. At best, you could split the code into different strings/"files" and concatenate them to kinda simulate multiple files.

We might need to look into that, because I'm suspicious of the tests as written 🤔. Do they all pass right now? I would expect the one that is checking for an invalid condition to not complain because nothing is saying what the tag is for the function being imported.

alexvy86 commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,91 @@
+		schema: [
+			{
+				type: "object",
+				properties: {
+					tags: {
+						type: "array",
+						items: { type: "string" },
+						uniqueItems: true,
+					},
+					exceptions: {
+						type: "array",
+						items: { type: "string" },
+						uniqueItems: true,
+					},
+				},
```
Should we make this into a dictionary with tags as keys and the list of exceptions for the given tag as values? So we can do something like "you can import @internal from ./foo, and @alpha from ./bar". Or can that be specified by using multiple entries in options? If so, then the functional logic of the rule probably cannot look at only `context.options[0]`, it needs to consider them all.

WayneFerrao replied:
Good idea, a map/dict would make it much more robust.

alexvy86 commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,91 @@
+						message: `Invalid tag format in rule configuration: '{${tag}}'. Tags should start with '@'.`,
+						data: { tag },
+					});
+					invalidTags = true;
```
`invalidTags` is not declared anywhere. Is setting it like this enough to cause a validation issue? Are there docs about it that we can link here?

WayneFerrao replied:
I think this was some erroneous code I left in.

alexvy86 reviewed (commented):
Haven't looked in detail at the meat of the validation, but leaving some other comments in the meantime.

WayneFerrao committed:
Refactor exceptions to a map, update code to simulate multiple files

WayneFerrao committed:
Refactored rule and code to use ts-morph. Updated tests to use distinct, local test files.

WayneFerrao committed:
Merge branch 'main' into preventImportsRule

WayneFerrao commented:
I've updated the rule to use ts-morph to traverse between files. In the current state,  the rule takes a look at the imports and visits the imported item's (function, interface etc) file to see if the block comment for that item contains a restricted tag. 
I've abstracted much of the logic to separate functions.

For more effective testing, including checks for exceptions and edge cases, I've introduced multiple local files in the test directory. The rule is configured to primarily utilize a tsconfig.json for file resolution, which aligns with the expected behavior for most packages in our repo. In scenarios where tsconfig.json is unavailable, the rule defaults to using the mockFiles folder in our test environment. To support this fallback mechanism, a tsconfig.json file has also been added to the test directory.

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+}
+
+/**
+ * Processes the tags array to ensure all tags are valid.
```
Nit: Might be a bit more accurate to say "filters the tags" and note that we will report any invalid ones we get.

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+ * Validate the exceptions object, ensuring all tags and paths are valid.
+ * @param exceptions - The exceptions object from the rule configuration.
+ * @param context - The context object provided by ESLint.
+ * @returns - An object with tags as keys and Sets of paths as values.
```
Nit: might be worth adding an example of the expected return structure.

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+		type: "problem",
+		docs: {
+			description:
+				"This rule restricts imports from specified tags or non-public APIs. This to prevent accidental dependencies on internal, unstable or undocumented parts of the codebase.",
```
I think the rule is a bit more general than this. I might say something like "Restricts imports based on one or more TSDoc tags they are annotated with. Can be used to enforce release tag rules against imports, etc."

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+				"This rule restricts imports from specified tags or non-public APIs. This to prevent accidental dependencies on internal, unstable or undocumented parts of the codebase.",
+			category: "Best Practices",
+		},
+		fixable: "code",
```
This tells eslint that violations of this rule can be auto-fixed. But I don't think that's true for our rule. I think we need to remove this?

WayneFerrao replied:
Good catch, will remove

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+			tsConfigPath = path.resolve(context.getCwd(), context.parserOptions.project);
+		} else {
+			// Fallback to a default/test path or handle absence of tsconfig.json as needed
+			tsConfigPath = path.join(__dirname, "../test/mockFiles/**/*.ts");
```
I don't think we want to ship this. We should probably just throw an exception (or whatever the correct way to notify eslint of a failure is) instead.

Josmithr replied:
It would probably be okay to attempt to fall back to "tsconfig.json", but beyond that we should throw.

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+		}
+		project.addSourceFilesAtPaths(tsConfigPath);
+
+		// const sourceFile = project.getSourceFileOrThrow("mockModule.ts");
```
Should this be removed?

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+			// Fallback to a default/test path or handle absence of tsconfig.json as needed
+			tsConfigPath = path.join(__dirname, "../test/mockFiles/**/*.ts");
+		}
+		project.addSourceFilesAtPaths(tsConfigPath);
```
Nit: it looks like we could move the instantiation of `project` down here and provide the tsconfigPath at construction.
https://ts-morph.com/setup/adding-source-files

Josmithr commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+
+				// For each item being imported
+				node.specifiers.forEach((specifier) => {
+					if (specifier.type === "ImportSpecifier") {
```
What do you think about moving this logic out into its own function?

Josmithr commented on common/build/eslint-config-fluid/package.json:
```diff
@@ -23,9 +24,12 @@
+		"test:mocha": "mocha --ignore \"dist/test/types/*\" --recursive dist/test -r node_modules/@fluidframework/mocha-test-setup"
 	},
 	"dependencies": {
+		"@microsoft/tsdoc": "^0.14.2",
```
Looks like we don't need this anymore :)

Josmithr reviewed (commented):
I left a few more comments, mostly nitpicks. This is looking really great! And the testing pattern you added looks great too!

WayneFerrao commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+ * @returns true if valid tag, else false.
+ */
+function isTagValid(tag, context) {
+	if (!tag.startsWith("@")) {
```
@Josmithr Checking for the @ is a good start but I guess we could also create a list of valid TSDoc tag names (found in the TSDoc documentation)  and check against that. I don't think we need a utility

Josmithr replied:
TSDoc tags are extensible (you can create custom ones), so we can't enforce a specific list. I was just curious if TSDoc had any other rules for what constitutes a valid tag / if their library had a utility for verifying. I think what you have here is fine - I was just curious.

WayneFerrao commented on common/build/eslint-config-fluid/src/custom-rules/no-restricted-tags-imports.js:
```diff
@@ -0,0 +1,161 @@
+ *
+ * @param importPath - The path used in the import statement.
+ * @param currentFilePath - The absolute path of the current file.
+ * @return - The absolute path of the imported module.
```
@Josmithr No, it's the absolute file path. The path relative from the current file to imported file is in `specifier.parent.source.value`. That has to be resolved with the current file's path in order to add it to the project and access it.

WayneFerrao committed:
PR fixes, update docs, add error for missing tsconfig

WayneFerrao commented on common/build/eslint-config-fluid/src/test/enforce-no-restricted-tags-imports.test.js:
```diff
@@ -0,0 +1,103 @@
+			},
+			parser: "@typescript-eslint/parser",
+			parserOptions: {
+				project: path.join(__dirname, "./tsconfig.json"),
```
Deciding to keep the tsconfig in tests/ as we discussed because this package doesn't use TS and moving it to root would likely cause confusion.

WayneFerrao committed:
Remove tsdoc

WayneFerrao committed:
Added comments

WayneFerrao committed:
fix policy check

WayneFerrao committed:
Added eslint-config-fluid to mocha exclusion

Josmithr reviewed (approved):
🚀

jikim-msft commented on common/build/eslint-config-fluid/src/test/mockFiles/mockModule.ts:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/**
+ * A mock function tagged with internal.
```
We should change this `public` right?

jikim-msft commented on common/build/eslint-config-fluid/src/test/mockFiles/fileWithExceptionImports.ts:
```diff
@@ -0,0 +1,6 @@
+ * Licensed under the MIT License.
+ */
+
+import { exceptionInternalFunction, alphaFunction2 } from "./exceptionFile.ts";
```
Why do we need this?

jikim-msft replied:
Resolved.
